### PR TITLE
[NativeAnimated][iOS] Add support for value listener

### PR DESCRIPTION
**Pull request by janicduplessis**
Adds support for `Animated.Value#addListener` for native driven animated values. Same as #8844 but for iOS. This depends on some JS code in #8844 so only review the 2nd commit and let's wait for #8844 to land first.

**Test plan**
Tested using the UIExplorer example.

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
By analyzing the blame information on this pull request, we identified @buba447 and @kmagiera to be potential reviewers.

**Comment by ryangomba** [comment]
@janicduplessis oh snap, thanks for doing this!

**Comment by ryangomba** [comment]
Circle is unhappy, but otherwise looks good to me. Will wait until #8844 is landed.

**Comment by janicduplessis** [comment]
Ok tests should pass now, NativeEventEmitter triggered an invariant because native modules are not defined during tests so I just mocked it.

**Comment by ide** [comment]
@janicduplessis can you rebase? #8844 has been merged too.

**Comment by facebook-github-bot-0** [comment]
@janicduplessis updated the pull request.

**Comment by janicduplessis** [comment]
@ide Done!

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by brentvatne** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-2** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3681749) to review internal test results.

**Comment by facebook-github-bot-4** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by facebook-github-bot-1** [comment]
@janicduplessis updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@janicduplessis updated the pull request.

**Comment by ide** [comment]
TravisCI iOS tests have been passing.. will try again

@facebook-github-bot shipit

**Comment by facebook-github-bot-7** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3681749) to review internal test results.

**Comment by facebook-github-bot-4** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by janicduplessis** [comment]
@javache Could you check why the import failed? There is no conflict and OSS tests all passed.

File: Libraries/Animated/src/NativeAnimatedHelper.js [hunk 1]
```diff
@@ -166,5 +164,5 @@ module.exports = {
   generateNewNodeTag,
   generateNewAnimationId,
   assertNativeAnimatedModule,
-  supportsNativeListener,
```
Comment (ide) [inline Libraries/Animated/src/NativeAnimatedHelper.js]: @janicduplessis just a guess, maybe FB is relying on this. We could stub it out as `return true` and try to merge.

Comment (janicduplessis) [inline Libraries/Animated/src/NativeAnimatedHelper.js]: I don't think so, I added it in the Android PR a few days ago.

**Comment by brentvatne** [comment]
@foghina maybe :)

**Comment by foghina** [comment]
Had a look but this is too iOS-y for me. A build seems to be failing with (I think) this error:

```
Invariant Violation: Native module cannot be null.
    at invariant (evalmachine.<anonymous>:2406:7)
    at new NativeEventEmitter (evalmachine.<anonymous>:4459:51)
    at evalmachine.<anonymous>:71282:24
    at loadModuleImplementation (evalmachine.<anonymous>:127:1)
    at guardedLoadModule (evalmachine.<anonymous>:70:8)
    at _require (evalmachine.<anonymous>:54:1)
    at evalmachine.<anonymous>:69201:23
    at loadModuleImplementation (evalmachine.<anonymous>:127:1)
    at guardedLoadModule (evalmachine.<anonymous>:70:8)
    at _require (evalmachine.<anonymous>:54:1)
```

Not sure if this is useful in any way. Leaving this up to @javache to figure out.

**Comment by janicduplessis** [comment]
I think I figured out why, some internal apps probably don't include the native `AnimatedModule` and this creates a `NativeEventEmitter` even if we don't use it. I'll make the `nativeEventEmitter` creation lazy that way it won't get created for apps that don't use it.

**Comment by ide** [comment]
Thanks @foghina, that stack trace is helpful. "evalmachine" is part of Node's vm module, perhaps this is failing in a Jest test?

**Comment by facebook-github-bot-2** [comment]
@janicduplessis updated the pull request.

**Comment by janicduplessis** [comment]
@ide Looks like it, I had to mock NativeEventEmitter in a test to make it pass but now I changed the NativeEventEmitter creation to be lazy so it won't need to be mocked unless the test uses an animated value listener for a native driven node.

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-8** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3681749) to review internal test results.

**Pull request closed without merging**